### PR TITLE
OpenAI wrappers for JS and PY

### PR DESCRIPTION
**Pull request by ankrgyl**
A really thin wrapper for the OpenAI SDK (v4) that:

* Preserves the type information of the original OpenAI SDK object
* Wraps the chat completion methods

Fixes BRA-559, BRA-560

### CODE REVIEW

**Comment by ankrgyl** [comment]
@manugoyal one thing I've been thinking about is whether we should wrap the Python module into a wrapper instead of a class (as I have). E.g. using [__getattr__](https://peps.python.org/pep-0562/) at the module level.

File: py/src/braintrust/oai.py [hunk 1]
```diff
@@ -1,41 +1,128 @@
...
+                                )
+                                first = False
+                            all_results.append(item)
+                            yield item
+                        span.log(output=all_results)
+                    finally:
+                        span.end()
 
+                should_end = False
+                return (x for x in gen())
```
Comment (manugoyal) [inline py/src/braintrust/oai.py:40]: does `return gen()` not work for some reason?

File: py/src/braintrust/oai.py [hunk 2]
```diff
@@ -1,41 +1,128 @@
...
+import time
 
-import json
-import sqlite3
+from .logger import current_span
 
-from .cache import CACHE_PATH
 
-OAI_CACHE = CACHE_PATH / "oai.sqlite"
+class ChatCompletionWrapper:
```
Comment (manugoyal) [inline py/src/braintrust/oai.py:6]: I wonder if this wrapper could be used for more than just `ChatCompletion`. What about any object that derives from `EngineAPIResource`?

**Review by manugoyal — COMMENTED** [review]
Cool. Only looked at python. Yeah I think wrapping the module could be cool. So that people just have to change their imports from `import openai` to `import braintrust_openai` or whatever?

File: py/src/braintrust/oai.py [hunk 3]
```diff
@@ -1,41 +1,128 @@
-# TODO MOVE TO AUTOEVALS IMPORT
-# https://github.com/braintrustdata/braintrust/issues/218
+import time
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply py/src/braintrust/oai.py:1]: done

File: py/src/braintrust/oai.py (hunk 2 above)
Comment (ankrgyl) [reply py/src/braintrust/oai.py:6]: Each one has distinct things we'd want to log I think

File: py/src/braintrust/oai.py (hunk 1 above)
Comment (ankrgyl) [reply py/src/braintrust/oai.py:40]: It does

File: py/src/braintrust/oai.py [hunk 4]
```diff
@@ -1,41 +1,128 @@
...
+            "input": messages,
+            "metadata": params,
+        }
 
-    return resp
+    def __getattr__(self, name):
+        return getattr(self.chat, name)
+
+
+class OpenAIWrapper:
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply py/src/braintrust/oai.py:113]: We should detect the various versions, and wrap them separately. I'm about to do this to handle the new version

File: py/src/braintrust/logger.py [hunk 5]
```diff
@@ -333,7 +333,7 @@ def construct_json_array(items):
 
 
 class _LogThread:
-    def __init__(self, name=None):
+    def __init__(self, name=None, lazy_login=None):
```
[the remark this answers is not in the document]
Comment (manugoyal) [reply py/src/braintrust/logger.py]: Can we add a unit test?

File: py/src/braintrust/logger.py [hunk 6]
```diff
@@ -791,13 +794,17 @@ def decorator(span_args, span_kwargs, f):
 
         @wraps(f)
         def wrapper_sync(*f_args, **f_kwargs):
-            with start_span(*span_args, **span_kwargs):
-                return f(*f_args, **f_kwargs)
+            with start_span(*span_args, **span_kwargs) as span:
```
Comment (manugoyal) [inline py/src/braintrust/logger.py:797]: Can we put these changes in a separate PR? I am worried about having substantially different behavior between the python and javascript SDKs, and I don't want to pollute this PR with that conversation.

File: py/src/braintrust/oai.py (hunk 2 above)
Comment (manugoyal) [reply py/src/braintrust/oai.py:14]: Is that true? I think both [`Completion`](https://platform.openai.com/docs/api-reference/completions/create) and [`Edit`](https://platform.openai.com/docs/api-reference/edits) have the same schema, as far as we are drilling down.

File: py/src/braintrust/oai.py [hunk 7]
```diff
@@ -1,41 +1,186 @@
...
+
+
+class ChatV1Wrapper(NamedWrapper):
+    def __init__(self, chat):
+        super().__init__(chat)
 
-def run_cached_request(Completion=None, **kwargs):
-    if Completion is None:
-        # OpenAI is very slow to import, so we only do it if we need it
         import openai
```
Comment (manugoyal) [inline py/src/braintrust/oai.py:162]: we don't need this import anymore?

File: py/src/braintrust/oai.py [hunk 8]
```diff
@@ -1,41 +1,186 @@
...
-    resp = Completion.create(**kwargs).to_dict()
 
-    cursor.execute("""INSERT INTO "cache" VALUES (?, ?)""", [param_key, json.dumps(resp)])
-    conn.commit()
+def wrap_openai(openai):
+    """
+    Wrap the openai module to add tracing. If Braintrust is not configured, this is a no-op.
 
-    return resp
+    :param openai: The openai module
```
Comment (manugoyal) [inline py/src/braintrust/oai.py:181]: For v1 they should be passing in the client object, not the module, right?

File: js/src/logger.ts [hunk 9]
```diff
@@ -1231,7 +1231,7 @@ function _urljoin(...parts: string[]): string {
   return parts.map((x) => x.replace(/^\//, "")).join("/");
 }
 
-function getCurrentUnixTimestamp(): number {
+export function getCurrentUnixTimestamp(): number {
```
Comment (manugoyal) [inline js/src/logger.ts:1234]: Could this go in `util`?

File: js/src/oai.ts [hunk 10]
```diff
@@ -1,24 +1,154 @@
...
+        prompt_tokens: number;
+        completion_tokens: number;
+      }
+    | undefined;
+}
+
+// TODO: Mock this up better
+type StreamingChatResponse = any;
+
+function wrapChatCompletionNonStreaming<
```
Comment (manugoyal) [inline js/src/oai.ts:77]: Should we just call this `wrapChatCompletion` if it also wraps streaming?

File: js/src/oai.ts [hunk 11]
```diff
@@ -1,24 +1,154 @@
...
+  P extends ChatParams,
+  C extends NonStreamingChatResponse | StreamingChatResponse
+>(completion: (params: P) => Promise<C>): (params: P) => Promise<any> {
+  return async (params: P) => {
+    if (params.stream) {
+      const { messages, ...rest } = params;
+      const span = currentSpan().startSpan("OpenAI Chat Completion", {
```
Comment (manugoyal) [inline js/src/oai.ts:84]: should we use the global method `startSpan`? `currentSpan` will not work if your active object is an experiment/logger. Also could we share the event content across both branches?
```
const { messages, ...rest } = params;
const initEvent = {
  input: messages,
  metadata: { ...rest }
};
...
```

File: js/src/oai.ts [hunk 12]
```diff
@@ -1,24 +1,154 @@
...
+        first = false;
+      }
+
+      allResults.push(item);
+      yield item;
+    }
+    this.span.log({
+      output: allResults,
+    });
+    this.span.end();
```
Comment (manugoyal) [inline js/src/oai.ts:152]: should we do try/finally here (as in python) just for safety?

**Review by manugoyal — COMMENTED** [review]
overall lgtm. Are there examples we can add to test this out and make sure things still work over time?

File: js/src/logger.ts (hunk 9 above)
Comment (ankrgyl) [reply js/src/logger.ts]: yep

File: js/src/oai.ts (hunk 10 above)
Comment (ankrgyl) [reply js/src/oai.ts]: yep

File: js/src/oai.ts (hunk 11 above)
Comment (ankrgyl) [reply js/src/oai.ts]: Yeah good call

File: js/src/oai.ts (hunk 12 above)
Comment (ankrgyl) [reply js/src/oai.ts]: Yep

File: py/src/braintrust/logger.py (hunk 6 above)
Comment (ankrgyl) [reply py/src/braintrust/logger.py]: Yep

Comment (ankrgyl) [reply py/src/braintrust/logger.py]: https://github.com/braintrustdata/braintrust-sdk/pull/33

File: py/src/braintrust/oai.py (hunk 2 above)
Comment (ankrgyl) [reply py/src/braintrust/oai.py:14]: I haven't looked at the return types -- in general i think we should proceed with caution until we thoroughly test.

File: py/src/braintrust/oai.py (hunk 7 above)
Comment (ankrgyl) [reply py/src/braintrust/oai.py:162]: we do on the RHS because we do a typecheck

File: py/src/braintrust/logger.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ankrgyl) [reply py/src/braintrust/logger.py]: added

**Review by manugoyal — APPROVED** [review]
